### PR TITLE
Improve blob path error messages

### LINKED ISSUE #1136: Improve Blob path validation
If you use the path separator the wrong way as in the following code:

```csharp
public static void BlobTrigger([BlobTrigger(@"test\{name}")] string blob, string name)
{
    Console.WriteLine($"Processed blob: {blob}");
}
```

You get the following runtime error, rather than an index time validation error. We should be catching this early.

```
Microsoft.WindowsAzure.Storage.StorageException occurred
  HResult=0x80131500
  Message=The remote server returned an error: (400) Bad Request.
  Source=Microsoft.WindowsAzure.Storage
  StackTrace:
   at Microsoft.WindowsAzure.Storage.Core.Executor.Executor.EndExecuteAsync[T](IAsyncResult result)
   at Microsoft.WindowsAzure.Storage.Blob.CloudBlobContainer.EndListBlobsSegmented(IAsyncResult asyncResult)
   at Microsoft.WindowsAzure.Storage.Core.Util.AsyncExtensions.<>c__DisplayClass1`1.<CreateCallback>b__0(IAsyncResult ar)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.WebJobs.Host.Storage.Blob.StorageBlobContainer.<ListBlobsSegmentedAsyncCore>d__20.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.WebJobs.Host.Blobs.Listeners.ScanBlobScanLogHybridPollingStrategy.<PollNewBlobsAsync>d__17.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.WebJobs.Host.Blobs.Listeners.ScanBlobScanLogHybridPollingStrategy.<PollAndNotify>d__13.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.WebJobs.Host.Blobs.Listeners.ScanBlobScanLogHybridPollingStrategy.<ExecuteAsync>d__12.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.WebJobs.Host.Timers.TaskSeriesTimer.<RunAsync>d__14.MoveNext()
   at Microsoft.Azure.WebJobs.Host.Timers.WebJobsExceptionHandler.<>c__DisplayClass3_0.<OnUnhandledExceptionAsync>b__0()
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()

Inner Exception 1:
WebException: The remote server returned an error: (400) Bad Request.
```

### PR DESCRIPTION
**Pull request by mamaso**
resolves #1136 
resolves https://github.com/Azure/azure-webjobs-sdk-script/issues/1482

### CODE REVIEW

File: test/Microsoft.Azure.WebJobs.Host.UnitTests/Blobs/BlobPathSourceTests.cs [hunk 1]
```diff
@@ -46,6 +46,27 @@ public void TestToString()
...
+
+        [Fact]
+        public void BlobPath_Throws_OnEmpty()
+        {
+            var exc = Assert.Throws<FormatException>(() => BlobPathSource.Create("/"));
+            Assert.Contains("Paths must be in the format 'container/blob'", exc.Message);
+        }
+
+        [Fact]
+        public void BlobPath_Throws_OnContainerResolves()
```
Comment (MikeStall) [inline test/Microsoft.Azure.WebJobs.Host.UnitTests/Blobs/BlobPathSourceTests.cs]: Is this true? 
I know we have this restriction for BlobTriggers, but is this applying it to regular blobs inputs as well?

Comment (mamaso) [inline test/Microsoft.Azure.WebJobs.Host.UnitTests/Blobs/BlobPathSourceTests.cs]: This only applies to triggers

File: src/Microsoft.Azure.WebJobs.Host/Blobs/BlobPathSource.cs [hunk 2]
```diff
@@ -18,28 +18,27 @@ public static IBlobPathSource Create(string pattern)
...
-
                 containerNamePattern = pattern.Substring(0, slashIndex);
                 blobNamePattern = pattern.Substring(slashIndex + 1);
             }
-            else
+
+            // There must be at least one character before the slash and one character after the slash.
+            bool hasNonEmptyBlobAndContainerNames = slashIndex > 0 && slashIndex < pattern.Length - 1;
+
+            if ((hasBlobName && !hasNonEmptyBlobAndContainerNames) || containerNamePattern.Contains('\\'))
```
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.Host/Blobs/BlobPathSource.cs]: >containerNamePattern.Contains('\\') [](start = 70, length = 35)

Why doesn't line 51 catch this?

Comment (mamaso) [inline src/Microsoft.Azure.WebJobs.Host/Blobs/BlobPathSource.cs]: If there are template.Parameter names we return a ParameterizedBlobPathSource without validating the container name.

One good point - I think that I can bump up the ValidateContainerName above the BindingTemplateSource call if we're disallowing tokens in the container name anyway

Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.Host/Blobs/BlobPathSource.cs]: Aren't { } allowed in contains for non-triggers? (And if this is a trigger-only path, does that mean we need to add the \ check to the non-trigger blob path?) 
Does your first sentence imply that that's another path we need to add this check to? 

---
In reply to: [118089716](https://github.com/Azure/azure-webjobs-sdk/pull/1155#discussion_r118089716) [](ancestors = 118089716)

Comment (mamaso) [inline src/Microsoft.Azure.WebJobs.Host/Blobs/BlobPathSource.cs]: Input/output bindings use `BlobPath` - you can see where I standardized the error above. They already handled a backslash.

**Pull request merged**